### PR TITLE
[spinel] support in place unpack

### PR DESCRIPTION
**Pull request by bukepo**
This RP added a new API `spinel_datatype_unpack_in_place ()` in spinel parser to support parsing composite data and filling in supplied buffers.

The current parser API returns data only valid inside the frame handler. This makes it hard to implement a general property getter like `GetProperty(spinel_key_t aKey, const char *aFormat, ...)` where the `aFormat` is the structure of the expected response. The newly added API accepts composite-type data pointer instead of pointer to composite-type data pointer, which would save repeated code for parsing and simplify host side implementing general getter for spinel properties.

### CODE REVIEW

File: src/ncp/spinel.c [hunk 1]
```diff
@@ -332,7 +331,14 @@ spinel_datatype_vunpack_(const uint8_t *data_ptr, spinel_size_t data_len, const
...
             {
-                *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                if (*arg_ptr)
+                {
+                    memcpy(*arg_ptr, data_ptr, sizeof(spinel_ipv6addr_t));
+                }
+                else
+                {
+                    *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                }
```
Comment (abtink) [inline src/ncp/spinel.c]: @bukepo I understand the optimization here and I agree it will helpful in some case, but I think it's not a good idea to change the behavior of the spinel C APIs for two reasons:

1) The `spinel.c` and `spinel.h` source files are directly integrated and used in other projects (e.g., in `wpantund` and some other internal projects). Changing this will have an impact on other projects.

2) While this makes the API more powerful, at the same time it becomes very error-prone and sort of dangerous. If the user/caller uses a local pointer variable (which is a very common use pattern) and forgets to initialize it to `NULL` then the data can be copied into some unknown/random memory region leading to memory corruption.  

I think it'd be safer to keep the C spinel APIs simple and safe and not change them.

I have been working on adding C++ wrapper classes over spinel C APIs (a `SpinelEncoder` and `SpinelDecoder`) to add more type-checking/safety to the code and help simplify `NcpBase` getter/setter methods. It's still work under progress (the `SpinelEncoder` code is in my OT fork in this (under-dev) branch https://github.com/abtink/openthread/commits/pr/spinel-encoder). We can consider adding support for such behavior in the C++ wrapper classes.

Comment (bukepo) [inline src/ncp/spinel.c]: Thanks for sharing your concerns. However, I think it's non-initialized pointers that is error-prone and dangerous. We should keep the practice to initialize a pointer to null if there's not a exact address for it, not only for the spinel API, but also other places.

This PR did have impact on other projects, that is, requiring pointers passing to spinel parser to be initialized to null. I don't think this impact is bad, indeed. And it doesn't take very much efforts to add initialization.

I did consider providing a wrapper before submitting this PR, however, I didn't figure out how to deal with the variable arguments. Is there a clean way to build a `va_list` from another? And if not using variable arguments, it will result in a lot of code. I like the simplicity of the spinel API and I prefer to keep it simple.

Thoughts? @abtink

Comment (abtink) [inline src/ncp/spinel.c]: I fully agree that it's a good practice to initialize all local variables (including pointers).

However, in the case of the use pattern for `spinel_datapack_unpack()` the fact that a pointer variable may be un-initialized is not really unsafe or error-prone:

- What we are sending to `unpack()` function is actually a pointer to a variable and we expect the `unpack()` function to update that variable. Basically the variable is used as an output of the `unpack()` function. For some simpler data types like `UINT16` the variable is a `uint16_t` but for more complex data types the variable happens to be a pointer to something. 

- If the `unpack()` is successful, the variable is updated and should then contain a valid value, so when the variable itself is a pointer, if the unpacking is successful it should be updated to point to what was unpacked. It should not matter if a pointer variable was initialized or not (the same way that it should not matter if the `uint16_t` variable was initialized). The key is to check for success of `unpack()` operation before using any of the output variables (including pointer types).

Your change in this PR makes the `unpack()` treat the case where the variable associated with the type is itself a pointer to something, differently than other simpler types. If the type is a pointer to something, with your change whether the variable is `NULL` or not, matters and changes the behavior (if it is `NULL` we update the variable, otherwise we copy some content at the address being pointed by variable and not change the variable itself).

The impact of this change on other projects using `spinel.c` is not limited to ensuring that pointer variables are initialized to `NULL` when they are declared.  For example, consider the code performing the unpacking of some table (say IPv6 address table, on-mesh prefixes, child table, whitelist, ...).  As we loop through and unpack each entry in the table, we typically reuse the same set of variables. With your changes to keep the behavior as before, we need to ensure to set all pointer type variables to `NULL` every time before calling the `unpack` (or any of its many derivatives). This requires some careful changes to all other projects and honestly I think it is something that can be easily missed now or used incorrectly in future. That's why I think the risks of this change far outweigh any possible benefit/flexibility it may provide.

Comment (bukepo) [inline src/ncp/spinel.c]: Thanks for the explanation, I agree that we should minimize the impact on other projects and lower the possibility of misuse.

For minimizing the impact, I think we can keep the behaviors unchanged for existing APIs. For the misuse concern, I'd like to provide a new C API to support this in-place with well documented notes about the usage. The new API and existing APIs would share most code, so that we don't bloat the code size. Thoughts?

Comment (abtink) [inline src/ncp/spinel.c]: :+1:  I like the new model of defining a new API. Thanks.

**Comment by darconeous** [comment]
A lot of thought went into the original design of `spinel_datatype_unpack()`. There are two design goals that the current mechanism satisfies:

* Argument symmetry with `spinel_datatype_pack()`. The only difference between the arguments you pass in to  `_unpack()` versus `_pack()` is that `_unpack()` takes a pointer to the type that would be passed in to `_pack()`. So a `int` for `_pack()` becomes a `int*` for `_unpack()`. A `char*` for `_pack()` becomes a `char**` for `_unpack()`. And, importantly, a `spinel_eui64_t*` for `_pack()` becomes a `spinel_eui64_t**` for `_unpack()`. This helps reduce argument confusion.
* Minimize unnecessary copies, to improve performance. If you need a copy of something outside of the context that the frame will exist, the intended solution is to make a copy of it after unpacking.

It is not terribly clear to me exactly what specific onerous problem this change is attempting to solve. Can you provide some more specific examples?

**Comment by bukepo** [comment]
I'm implementing a general spinel property getter function with prototype like `GetProperty(spinel_key_t aKey, const char *aFormat, ...)`.
The caller assumes all data are properly assigned when the function returns successfully. Instead of writing parsing code for each property, I reuse the `spinel_datatype_vunpack*()` API to implement a simple general property handler for all properties with the supplied `aFormat` as the `pack_format` argument of `spinel_datatype_vunpack*()`. Since composite data like string returned from `spinel_datatype_vunpack()` are actually located in the HDLC frame, which are only valid during the HDLC decoder's frame callback. Once `GetProperty(spinel_key_t aKey, const char *aFormat, ...)` returned, these data already became invalid. So in this PR I implemented a alternative way for parsing the spinel data so that composite data are copied into the caller supplied buffers.

As for the symmetry design, it's really beautiful and I like it. And I also noticed that `scanf()` and `printf()` are very similar to spinel's `*unpack()` and `*pack()`, reader and writer. The reader `scanf()` just like `*unpack()` accepts a pointer to the type that would be passed in to the writer `printf()`, while for string both `scanf()` and `printf()` accepts a pointer to `char` and the format code is the same `%s`. I didn't feel confused, IMO, maybe because in C, a parameter is passed by pointer if,
1. the parameter is going to be changed by the function,
2. or the parameter is composite data types and passing by pointer is more efficient by value.
So maybe, it's acceptable for composite data type that they are not that symmetry in reader and writer, i.e. `*unpack()` and `*pack()`.

As for the performance concern, this PR didn't change the existing API, so the caller only use the in place parsing API if needed. Besides, to enable making copy after unpacking as suggested, that lead to providing specified parsing code for each spinel property, which may result in a lot of repeated code.

Thoughts? @darconeous

**Comment by darconeous** [comment]
Can you paste in some code showing me how you are using this new API to make things nicer? I think that would make it's utility more clear to me.

**Comment by bukepo** [comment]
```cpp
static va_list sArgs;
static const char* sFormat;

void GetEui64(uint8_t *aEui64)
{
    GetProp(SPINEL_PROP_HWADDR, SPINEL_DATATYPE_EUI64_S, aEui64);
}

int GetProp(spinel_key_t aKey, const char* aFormat, ...);
{
    sFormat = aFormat;
    va_start(sArgs, aFormat);
    int ret = SendGetPropAndWaitReply(aKey);
    va_end(sArgs);
    return ret;
}

int SendGetPropAndWaitReply(spinel_key_t aKey)
{
    // code to send  the get property
    ...
    // loop waiting for serial port data and feed to HDLC decoder.
   char buf[];
    while (/* the property response is not received or not timeout */)
    {
        int rval = read(sSock, buf, sizeof(buf));
        sHdlcDecoder.Decode(buf, rval);
    }
    return error;
}

void HandleHdlcFrame(uint8_t *aBuffer, uint16_t aBufferLength)
{
    // some code to parse command
    ...
    // if this is response of the get prop command, aBuffer has already been adjusted to point to the start of the property value.
    {
        rval = spinel_datatype_vunpack_in_place(aBuffer, aBufferLenth, aFormat, aArgs);
    }
    ...
}
```

**Comment by abtink** [comment]
Thinking about this, handling the failure case  may require some care with the new model specially for handling of structs and types that contain more than one value.

When the `spinel_datatype_vunpack_in_place` is called we don't yet know if the received spinel frame (content of `aBuffer`) is valid and can be parsed correctly for the given format. We need to ensure that for a failed case, the passed-in variables remain unchanged.  This become tricky if the format contains structs or more than one value, as the `unpack` may be able to parse/decode some of the values but not all.

The existing model/API addresses this as follows:
- The caller upacks the content into local variables (gets pointers into the buffer where the data is without copying it over),
- Checks the status of upack and ensure the operation succeeded 
- Only if successful, the caller copies/moves the content where they need to go.

Wit new model, we may need to keep a copy of original values of parameters sent to `int GetProp(spinel_key_t aKey, const char* aFormat, ...);` to be able to revert back in case of failure...

**Comment by darconeous** [comment]
Thanks @bukepo, that helps my understanding greatly. I see you are trying to write blocking API calls that combine both sending the request and parsing the response. Introducing this method makes for much more elegant code, so I can see your motivation.

I'm going to give a little more thought to this and respond more decisively later today.

Regarding @abtink's comment:

> We need to ensure that for a failed case, the passed-in variables remain unchanged.

This was never a guarantee, even with the standard `spinel_datatype_vunpack()`. All arguments that were successfully parsed up to the point of the parse error will be changed. This generally isn't a problem as long as you can signal some sort of error condition that indicates that the frame was not parsed as expected. Unless I am missing something, I don't think there is a need to restore previous values.

**Comment by abtink** [comment]
@darconeous I meant from from caller of `GetProp(spinel_key_t aKey, const char* aFormat, ...);` perspective.

File: src/ncp/spinel.c [hunk 2]
```diff
@@ -45,7 +45,6 @@
 
 #include "spinel.h"
 
-#include <assert.h>
```
Comment (darconeous) [inline src/ncp/spinel.c]: Why is this being removed?

File: src/ncp/spinel.c [hunk 3]
```diff
@@ -327,12 +326,22 @@ spinel_datatype_vunpack_(const uint8_t *data_ptr, spinel_size_t data_len, const
 
         case SPINEL_DATATYPE_IPv6ADDR_C:
         {
-            spinel_ipv6addr_t **arg_ptr = va_arg(args->obj, spinel_ipv6addr_t **);
             require_action(data_len >= sizeof(spinel_ipv6addr_t), bail, (ret = -1, errno = EOVERFLOW));
 
-            if (arg_ptr)
+            if (in_place) {
```
Comment (darconeous) [inline src/ncp/spinel.c]: OpenThread style is (sadly) curly-braces go on their own lines.

File: src/ncp/spinel.h [hunk 4]
```diff
@@ -1455,8 +1460,47 @@ SPINEL_API_EXTERN spinel_ssize_t spinel_datatype_vpack(uint8_t *data_out, spinel
...
+ * This function parses spinel data similar to vsscanf().
+ *
+ * @param[in]   data_in     A pointer to the data to parse.
+ * @param[in]   data_len    The length of @p data_in in bytes.
+ * @param[in]   pack_format C string that contains a format string follows the same specification of spinel.
+ * @param[in]   args        A value identifying a variable arguments list.
+ *
+ * @returns The parsed length in bytes.
+ *
+ * @note This function behaves different from `spinel_datatype_vunpack()`:
```
Comment (darconeous) [inline src/ncp/spinel.h]: I would add this comment to `spinel_datatype_unpack_in_place()` as well.

File: tests/unit/test_address_sanitizer.cpp [hunk 5]
```diff
@@ -39,7 +39,7 @@ int main(int argc, char *argv[])
 
     for (int i = 0; i < 100; i++)
     {
-	stack_array[i] = i;
+        stack_array[i] = i;
```
Comment (darconeous) [inline tests/unit/test_address_sanitizer.cpp]: This change seems unrelated.

**Comment by darconeous** [comment]
Any chance I could get you to add unit tests (see bottom of file) that exercise the newly added code?

File: src/ncp/spinel.c (hunk 2 above)
Comment (bukepo) [inline src/ncp/spinel.c]: I didn't find where it is used in this file.

File: src/ncp/spinel.c (hunk 3 above)
Comment (bukepo) [inline src/ncp/spinel.c]: Thanks, I will update.

File: src/ncp/spinel.h (hunk 4 above)
Comment (bukepo) [inline src/ncp/spinel.h]: thanks, I've added.

**Comment by bukepo** [comment]
@darconeous I have added some unit tests.
